[BEAM-5807] Conversion from AVRO records to rows

kanterov opened the pull request:
This PR introduces an initial implementation of conversions from AVRO `GenericRecord` to `Row`. Transformation doesn't allow any type widening or narrowing. The idea is to build narrowing and widening as a separate transformation using `Row` and keep this code simple.

There are further steps, that I want to put into separate pull requests:
- change reading in `BigQueryTableProvider` to newer implementation
- the opposite, `Row` to `GenericRecord`, and test that roundtrip preserves equality
- add transforms for projecting fields, type widening and narrowing
- add a builder to `AvroIO`
- add `AvroTableProvider` for SQL

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

kanterov commented:
@akedin I would need one more day to have a fresh look, can you please take a look and give feedback on the approach?

akedin commented:
run java precommit

kanterov commented on sdks/java/core/src/test/java/org/apache/beam/sdk/util/AvroUtilsTest.java:
```diff
@@ -0,0 +1,114 @@
+    List<GenericRecord> records = Lists.newArrayList((Iterable<GenericRecord>) iterable);
+
+    for (GenericRecord record : records) {
+      AvroUtils.toRowStrict(record, schema);
```
I'm thinking of adding assertions when the `Row => GenericData` is ready. Assertion would be `fromRow(toRow(x)) == x`

As well as few unit tests in a different PR.

akedin replied:
makes sense

kanterov commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java:
```diff
@@ -381,7 +381,7 @@ public static Builder withSchema(Schema schema) {
     }
 
-    public Builder addValue(Object values) {
+    public Builder addValue(@Nullable Object values) {
```
I added annotation because findbugs complained, it's already widely used with nullable values, just wasn't caught before, meaning there would be only one usage of `addNullValue`

kanterov commented on sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/AvroUtils.java:
```diff
@@ -0,0 +1,349 @@
+
+  private static Object convertIntStrict(Integer value, Schema.FieldType fieldType) {
+    switch (fieldType.getTypeName()) {
+      case INT32:
```
Thanks for the suggestion. True! I've started with doing implicit casting and then removed this code because it was too complicated, now I will change to guards because there is no need in switch anymore.

kanterov commented on sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/AvroUtils.java:
```diff
@@ -0,0 +1,349 @@
+
+      // optional fields in AVRO have form of:
+      // {"name": "foo", "type": ["null", "something"]}
+      if (types.size() == 2 && types.get(0).equals(nullSchema)) {
```
True, I was assuming `null` should be always the first parameter, but it isn't true. Just checked out how it's done in Spark:
https://github.com/databricks/spark-avro/blob/branch-4.0/src/main/scala/com/databricks/spark/avro/SchemaConverters.scala#L76

Apparently, unions can also be nullable

kanterov commented:
@akedin Thanks for the review. I would be happy to merge this sooner than later.

I see this one of building blocks, that can be used to implement different higher-level transforms. I will provide a few examples I have in mind.

## Case 1. AvroIO and POJO

As I understand, this is what you mentioned. Given AVRO schema and POJO class, we want to read `PCollection<POJO>` using `AvroIO`.

How it works:
1. we get `Schema` for AVRO using AVRO schema
2. we get `Schema` for POJO from `SchemaProvider`
3. there is a `PTransform` that converts rows between schemas if needed
4. `PTransform#validate` checks if conversion is possible

No codegen is required as opposed to using avro-compiler.

## Case 2. AvroIO and Row

The difference from case 1 is that we don't have POJO. AVRO schema is dynamic, and we don't know it in compile time. We get it in runtime from external schema registry service, or from AVRO file headers in a filesystem. At the same time, we want to run SQL, the query itself can also be dynamic, or be based on a schema. For instance, you can imagine a generic data profiling pipeline implemented this way.

## Case 3. BigQuery and TableProvider

We get BigQuery table schema, and we know table schema, then it works similar to case 2. It is something like external tables in Hive.

We don't need to register tables in advance as it's done today, because we can list tables in a GCP project, and derive schemas for them. The same approach can be used for AVRO if there is metadata catalog like Hive metastore.

## Reading directly to Row

I was trying to read directly to `Row` instead of `GenericRecord`. It's possible to get up to 3 times performance improvement by doing runtime codegen using bytebuddy or similar. There is an [article](https://techblog.rtbhouse.com/2017/04/18/fast-avro/) describing the similar approach.`GenericRecord`. 

This approach enables projection and filter pushdowns to a data source. Recently there was another [article](https://dawn.cs.stanford.edu/2018/08/07/sparser/) about it. This could be done by injecting custom `DatumReader` into `AvroIO`.

akedin commented:
retest this please

akedin commented:
build fails because *.java files don't have Apache license headers, can you add them (copy-paste from existing java files)?

kanterov commented:
@akedin sure, I did one more walkthrough and addressed your comments.

I submitted a PR to add license headers with spotless, let me know what do you think:
- https://github.com/apache/beam/pull/6791

akedin commented:
Seems good now, can you rebase so that the tests pass after Kenn has made them deterministic?

akedin commented:
now spotless fails :)

kanterov commented:
Run Java FnApi PreCommit

akedin commented:
Looks like our build is especially broken in the last few days :( Next problem problem: https://github.com/apache/beam/pull/6807

kanterov commented:
@akedin it looks like tests pass :), two failing jobs aren't implemented yet, and being developed in separate PRs

kanterov commented:
@reuvenlax I'm not sure I understand what it means, can you elaborate?
